### ISSUE TITLE
oracle.to_date with Julian Days format returns wrong results

### ISSUE THREAD

**Issue by darold** [orafce/orafce#272]
Hi Pavel,

I have found an issue with the 'J' format of the `oracle.to_date(text, text)` function, it reports wrong results compared to Oracle.

On Oracle:
```
SQL> ALTER SESSION SET NLS_DATE_FORMAT='YYYY-MM-DD';

Session altered.

SQL> SELECT TO_DATE(098185, 'J') FROM DUAL;

TO_DATE(09
----------
4444-10-25
```
On PostgreSQL with Orafce:
```
gilles=# SELECT oracle.to_date('098185', 'J');
        to_date
------------------------
 4445-09-20 00:00:00 BC
```
This is due to the different referential date used by both RDBMS about the 'J' formatting:

- Oracle => Julian day; the number of days since January 1, 4712 BC.
- PostgreSQL => Julian Date (integer days since November 24, 4714 BC at local midnight;

I propose to modify the function `oracle.to_date(text, text)` as follow:
```
CREATE OR REPLACE FUNCTION oracle.to_date (TEXT, TEXT) RETURNS oracle.date AS $$
SELECT CASE WHEN upper($2) = 'J' THEN
                to_date($1, $2) + '430 days'::interval
        ELSE
                TO_TIMESTAMP($1,$2)::oracle.date
        END;
$$ LANGUAGE SQL
STRICT IMMUTABLE PARALLEL SAFE;
```
After this change the result correspond to what Oracle returns:
```
pgweb=# SELECT oracle.TO_DATE('098185', 'J');
        to_date         
------------------------
 4444-11-24 00:00:00 BC
```
If you agree to include this fix I can send a PR.

Cheers.

**Comment by okbob** [comment]
út 27. 8. 2024 v 22:51 odesílatel Gilles Darold ***@***.***>
napsal:

> Hi Pavel,
>
> I have found an issue with the 'J' format of the oracle.to_date(text,
> text) function, it reports wrong results compared to Oracle.
>
> On Oracle:
>
> SQL> ALTER SESSION SET NLS_DATE_FORMAT='YYYY-MM-DD';
>
> Session altered.
>
> SQL> SELECT TO_DATE(098185, 'J') FROM DUAL;
>
> TO_DATE(09
> ----------
> 4444-10-25
>
> On PostgreSQL with Orafce:
>
> gilles=# SELECT oracle.to_date('098185', 'J');
>         to_date
> ------------------------
>  4445-09-20 00:00:00 BC
>
> This is due to the different referential date used by both RDBMS about the
> 'J' formatting:
>
>    - Oracle => Julian day; the number of days since January 1, 4712 BC.
>    - PostgreSQL => Julian Date (integer days since November 24, 4714 BC
>    at local midnight;
>
> I propose to modify the function oracle.to_date(text, text) as follow:
>
> CREATE OR REPLACE FUNCTION oracle.to_date (TEXT, TEXT) RETURNS oracle.date AS $$
> SELECT CASE WHEN upper($2) = 'J' THEN
>                 to_date($1, $2) + '430 days'::interval
>         ELSE
>                 TO_TIMESTAMP($1,$2)::oracle.date
>         END;
> $$ LANGUAGE SQL
> STRICT IMMUTABLE PARALLEL SAFE;
>
> After this change the result correspond to what Oracle returns:
>
> pgweb=# SELECT oracle.TO_DATE('098185', 'J');
>         to_date
> ------------------------
>  4444-11-24 00:00:00 BC
>
> If you agree to include this fix I can send a PR.
>

ok

please send PR

**Comment by darold** [comment]
@okbob please wait a little, I have made a typo in the function, this is +400, not +430

**Comment by darold** [comment]
I have also forgotten to add a regression test :-(

**Comment by darold** [comment]
Please @okbob can you recall me how to execute the regression tests?

**Comment by okbob** [comment]
you need to have postgres devel packages, and pg_config in PATH. Then `make all; sudo make install; make installcheck` in orafce directory

**Comment by darold** [comment]
Ok, maybe I have made another typo too as I have tried make installcheck without success. Running now.

**Comment by darold** [comment]
PR pushed.

There is still a difference with Oracle. PostgreSQL adds BC to the date returned;
```
=# SELECT oracle.to_date('115032', 'j');
        to_date         
------------------------
 4398-12-10 00:00:00 BC
(1 row)
```
where Oracle returns only the date:
```
SQL> ALTER SESSION SET NLS_DATE_FORMAT='YYYY-MM-DD HH24:MI:SS';

Session altered.

SQL> SELECT to_date('115032', 'j') FROM DUAL;

TO_DATE('115032','J
-------------------
4398-12-10 00:00:00
```
I have some doubt if we should fix that too, maybe not.

**Comment by okbob** [comment]
st 28. 8. 2024 v 21:42 odesílatel Gilles Darold ***@***.***>
napsal:

> PR pushed.
>
> There is still a difference with Oracle. PostgreSQL adds BC to the date
> returned;
>
> =# SELECT oracle.to_date('115032', 'j');
>         to_date
> ------------------------
>  4398-12-10 00:00:00 BC
> (1 row)
>
> where Oracle returns only the date:
>
> SQL> ALTER SESSION SET NLS_DATE_FORMAT='YYYY-MM-DD HH24:MI:SS';
>
> Session altered.
>
> SQL> SELECT to_date('115032', 'j') FROM DUAL;
>
> TO_DATE('115032','J
> -------------------
> 4398-12-10 00:00:00
>
> I have some doubt if we should fix that too, maybe not.
>
the fix of this needs rewriting to_date(text,text) to orafce, and it is not
too nice. I have not any related use case, so you should to do decision
what is enough solution.

**Comment by darold** [comment]
Hi Pavel, I will send another PR to remove the BC part, actually we need the exact same output as Oracle, especially when the date is used as a string to be passed to some other procedures.

**Comment by okbob** [comment]
ok, please, do it

**Comment by darold** [comment]
PR sent

**Comment by okbob** [comment]
merged

do you plan to do some other changes, or I can release it?

pá 30. 8. 2024 v 7:52 odesílatel Gilles Darold ***@***.***>
napsal:

**Comment by okbob** [comment]
released

pá 30. 8. 2024 v 8:14 odesílatel Gilles Darold ***@***.***>
napsal:

**Comment by darold** [comment]
After some more investigations, for some reason that I can not explain, Oracle returns the same date than PostgreSQL if the date is greater or equal to `0200-03-01 00:00:00` but a différent result with a date bellow.

PostgreSQL:
```
gilles=# SELECT TO_NUMBER(TO_CHAR('0200-02-28'::date, 'J')) AS julian_number;
 julian_number 
---------------
       1794167
(1 row)

gilles=# SELECT to_date('1794167', 'j');
  to_date   
------------
 0200-02-28

(1 row)
```
In Oracle:
```
SQL> SELECT to_date('1794167', 'j') AS jd FROM DUAL;

 JD
-------------------
0200-02-29 00:00:00
```
and starting from this error any integer passed below this date will reports a wrong result compared to PostgreSQL.

Maybe we could fix the oracle.to_char(text,text) to issue a warning when the value passed is below this date?

**Comment by darold** [comment]
This doesn't explain why with two different date as JD referential for to_char(integer, 'J') we have the same result after this date. As I don't want to become crazy I stop the investigation here. If someone can explain the Oracle logic for this calculation I would be happy to know.

**Comment by okbob** [comment]
pá 30. 8. 2024 v 15:22 odesílatel Gilles Darold ***@***.***>
napsal:

> This doesn't explain why with two different date as JD referential for
> to_char(integer, 'J') we have the same result after this date. As I don't
> want to become crazy I stop the investigation here. If someone can explain
> the Oracle logic for this calculation I would be happy to know.
>

probably nobody uses this feature, just you. Please do investigation  in
this weekend if it is possible, and send PR of fix please.

I remember, when I tested orafce 15 years ago, and I tested date functions,
then I found a bug in Oracle. There is a warning in readme still

Built-in Oracle date functions have been tested against Oracle 10 for
conformance. Date ranges from 1960 to 2070 work correctly. Dates before
1100-03-01 cannot be verified due to a bug in Oracle.

But it was tested against Oracle 10.

**Comment by darold** [comment]
Ok, I will still look deeper into that.

**Comment by darold** [comment]
@okbob I have run a script to check the `to_date(..., 'J')` day bye day since now until the date returned by Oracle and PostgreSQL differs. The result:
```
Oracle : 1582-10-04 (2299160)
PostgreSQL : 1582-10-14 (2299160)
```
What I suggest is to use the PostgreSQL to_date function and raise an error when the parameter is <= 2299160 (1582-10-14).
For dates below we will possibly return a false result so I think we should raise an error. I can fix previous patch in this way if you agree.

**Comment by darold** [comment]
I forgot to say that I that tested against a 11g and 21c with the failing date.

**Comment by okbob** [comment]
st 4. 9. 2024 v 20:59 odesílatel Gilles Darold ***@***.***>
napsal:

> @okbob <https://github.com/okbob> I have run a script to check the to_date(...,
> 'J') day bye day since now until the date returned by Oracle and
> PostgreSQL differs. The result:
>
> Oracle : 1582-10-04 (2299160)
> PostgreSQL : 1582-10-14 (2299160)
>
> What I suggest is to use the PostgreSQL to_date function and raise an
> error when the parameter is <= 2299160 (1582-10-14).
> For dates below we will possibly return a false result so I think we
> should raise an error. I can fix previous patch in this way if you agree.
>
Maybe we can use some GUC like orafce.oracle_compatibility_date_limit

if you don't use J, then safe limit is 1100-03-01

default for this GUC can be 1582-10-04

**Comment by darold** [comment]
I agree, I will fix that in that way.

**Comment by okbob** [comment]
st 4. 9. 2024 v 22:03 odesílatel Gilles Darold ***@***.***>
napsal:

> I agree, I will fix that in that way.
>

Maybe orafce.oracle_compatibility_date_limit should be boolean, when it is
enabled, then check 1582-10-14 for J or 1100-03-01 for others, and raises
an exception. When it is false, then this check is disabled.

**Comment by darold** [comment]
What do you think the default value for the GUC should be? I think it should be enabled by default.

**Comment by okbob** [comment]
I opened orafce 4.13 - please, push your fix to this release (don't create new release)

**Comment by darold** [comment]
ok, I will do that today

**Comment by okbob** [comment]
čt 19. 9. 2024 v 6:48 odesílatel Gilles Darold ***@***.***>
napsal:

> ok, I will do that today
>
thank you

**Comment by darold** [comment]
Hi @okbob , PR #277 sent. Sorry for the delay, I've started to write it in plpgsql and decide to switch to a C implementation for speed.

**Comment by darold** [comment]
Documentation updated and PR sent. Please fix the documentation formatting for the to_date(text,text) function description. I don't know why the list item is indented and how to fix it.

**Comment by okbob** [comment]
released

**Comment by okbob** [comment]
> Documentation updated and PR sent. Please fix the documentation formatting for the to_date(text,text) function description. I don't know why the list item is indented and how to fix it.

fixed

**Comment by okbob** [comment]
ne 22. 9. 2024 v 21:03 odesílatel Gilles Darold ***@***.***>
napsal:

> Closed #272 <https://github.com/orafce/orafce/issues/272> as completed.
>

thank you for your work
